File: src/main/java/io/pageobject/generator/locator/protractor/ExpressionTextLocatorSource.java
package io.pageobject.generator.locator.protractor;

import io.pageobject.generator.GeneratorContext;
import io.pageobject.generator.locator.AbstractLocatorSource;
import org.jsoup.nodes.Element;

import static io.pageobject.generator.Expressions.getFirstExpression;

public class ExpressionTextLocatorSource extends AbstractLocatorSource {

    @Override
    public String extractLocatorValue(Element element, GeneratorContext context) {
        return getFirstExpression(element.ownText());
    }

    public String locatorTemplate() {
        return "by.exactBinding('%s')";
    }

}

File: lib/locator/protractor/expression-text-locator-source.js
"use strict";

const AbstractLocatorSource = require('../abstract-locator-source');
const _ = require('lodash');
const getFirstExpression = require('../../expressions').getFirstExpression;
const ownText = require('../../dom-helpers').ownText;

const template = _.template("by.exactBinding('<%= value %>')");

class ExpressionTextLocatorSource extends AbstractLocatorSource {

  extractLocatorValue(context) {
    return getFirstExpression(ownText(context.element));
  }

  locatorTemplate() {
    return template;
  }

}

module.exports = ExpressionTextLocatorSource;